File: examples/ios/Examples/ObjectModels.m
// :replace-start: {
//   "terms": {
//     "ObjectModelsExamplesObjc_": "",
//     "OptionalRequiredPropertyObjcExample_": ""
//   }
// }
#import <XCTest/XCTest.h>
#import <Realm/Realm.h>

// :code-block-start: array-declaration
// ObjectModelsExamplesObjc_Task.h
@interface ObjectModelsExamplesObjc_Task : RLMObject
@property NSString *description;
@end

// Define an RLMArray<ObjectModelsExamplesObjc_Task> type
RLM_ARRAY_TYPE(ObjectModelsExamplesObjc_Task)


// ObjectModelsExamplesObjc_User.h
// #include "ObjectModelsExamplesObjc_Task.h"
@interface ObjectModelsExamplesObjc_User : RLMObject
@property NSString *name;
// Use RLMArray<ObjectModelsExamplesObjc_Task> to have a list of tasks
// Note the required double tag (<ObjectModelsExamplesObjc_Task *><ObjectModelsExamplesObjc_Task>)
@property RLMArray<ObjectModelsExamplesObjc_Task *><ObjectModelsExamplesObjc_Task> *tasks;
@end
// :remove-start:
// ObjectModelsExamplesObjc_Task.m
@implementation ObjectModelsExamplesObjc_Task
@end

// ObjectModelsExamplesObjc_User.m
@implementation ObjectModelsExamplesObjc_User
@end
// :remove-end:
// :code-block-end:

// :code-block-start: define-a-model
// A dog has an _id primary key, a string name, an optional
// string breed, and a date of birth.
@interface ObjectModelsExamplesObjc_Dog : RLMObject
@property RLMObjectId *_id;
@property NSString *name;
@property NSString *breed;
@property NSDate *dateOfBirth;
@end

@implementation ObjectModelsExamplesObjc_Dog
+ (NSString *)primaryKey {
    return @"_id";
}

+ (NSArray<NSString *> *)requiredProperties {
    return @[
        @"_id", @"name", @"dateOfBirth"
    ];
}
@end
// :code-block-end:

// :code-block-start: optional-required-properties
@interface OptionalRequiredPropertyObjcExample_Person : RLMObject
// Required property - included in `requiredProperties`
// return value array
@property NSString *name;

// Optional string property - not included in `requiredProperties`
@property NSString *address;

// Required numeric property
@property int ageYears;

// Optional numeric properties use NSNumber tagged
// with RLMInt, RLMFloat, etc.
@property NSNumber<RLMFloat> *heightCm;
@end

@implementation OptionalRequiredPropertyObjcExample_Person
// Specify required pointer-type properties here.
// Implicitly required properties (such as properties
// of primitive types) do not need to be named here.
+ (NSArray<NSString *> *)requiredProperties {
    return @[@"name"];
}
@end
// :code-block-end:

// :code-block-start: specify-a-primary-key
@interface ObjectModelsExamplesObjc_Project : RLMObject
@property NSInteger id; // Intended primary key
@property NSString *name;
@end

@implementation ObjectModelsExamplesObjc_Project
// Return the name of the primary key property
+ (NSString *)primaryKey {
    return @"id";
}
@end
// :code-block-end:

// :code-block-start: index-a-property
@interface ObjectModelsExamplesObjc_Book : RLMObject
@property int priceCents;
@property NSString *title;
@end

@implementation ObjectModelsExamplesObjc_Book
// Return a list of indexed property names
+ (NSArray *)indexedProperties {
    return @[@"title"];
}
@end
// :code-block-end:

// :code-block-start: ignore-a-property
@interface ObjectModelsExamplesObjc_Person : RLMObject
@property NSInteger tmpId;
@property (readonly) NSString *name; // read-only properties are automatically ignored
@property NSString *firstName;
@property NSString *lastName;
@end

@implementation ObjectModelsExamplesObjc_Person
+ (NSArray *)ignoredProperties {
    return @[@"tmpId"];
}
- (NSString *)name {
    return [NSString stringWithFormat:@"%@ %@", self.firstName, self.lastName];
}
@end
// :code-block-end:

@interface ObjectModelsObjc : XCTestCase
@end

@implementation ObjectModelsObjc
@end

// :replace-end:

File: examples/ios/Examples/ObjectModels.swift
// :replace-start: {
//   "terms": {
//     "ObjectModelsExamples_": "",
//     "ObjectModelsObjcDynamicExamples_": "",
//     "OptionalRequiredPropertyExample_": "",
//     "OptionalRequiredPropertyObjcDynamicExample_": ""
//   }
// }
import XCTest
import RealmSwift

// :code-block-start: define-a-model
// A dog has an _id primary key, a string name, an optional
// string breed, and a date of birth.
class ObjectModelsExamples_Dog: Object {
    @Persisted(primaryKey: true) var _id: ObjectId
    @Persisted var name = ""
    @Persisted var breed: String?
    @Persisted var dateOfBirth = Date()
}
// :code-block-end:

// :code-block-start: optional-required-properties
class OptionalRequiredPropertyExample_Person: Object {
    // Required string property
    @Persisted var name = ""

    // Optional string property
    @Persisted var address: String?

    // Required numeric property
    @Persisted var ageYears = 0

    // Optional numeric property
    @Persisted var heightCm: Float?
}
// :code-block-end:

// :code-block-start: optional-required-properties-objc-dynamic
class OptionalRequiredPropertyObjcDynamicExample_Person: Object {
    // Required string property
    @objc dynamic var name = ""

    // Optional string property
    @objc dynamic var address: String?

    // Required numeric property
    @objc dynamic var ageYears = 0

    // Optional numeric property
    let heightCm = RealmProperty<Float?>()
}
// :code-block-end:

// :code-block-start: specify-a-primary-key
class ObjectModelsExamples_Project: Object {
    @Persisted(primaryKey: true) var id = 0
    @Persisted var name = ""
}
// :code-block-end:

// :code-block-start: specify-a-primary-key-objc-dynamic
class ObjectModelsObjcDynamicExamples_Project: Object {
    @objc dynamic var id = 0
    @objc dynamic var name = ""

    // Return the name of the primary key property
    override static func primaryKey() -> String? {
        return "id"
    }
}
// :code-block-end:

// :code-block-start: index-a-property
class ObjectModelsExamples_Book: Object {
    @Persisted var priceCents = 0
    @Persisted(indexed: true) var title = ""
}
// :code-block-end:

// :code-block-start: index-a-property-objc-dynamic
class ObjectModelsObjcDynamicExamples_Book: Object {
    @objc dynamic var priceCents = 0
    @objc dynamic var title = ""

    // Return a list of indexed property names
    override static func indexedProperties() -> [String] {
        return ["title"]
    }
}
// :code-block-end:

// :code-block-start: ignore-a-property
class ObjectModelsExamples_Person: Object {
    // If some properties are marked as @Persisted,
    // any properties that do not have the @Persisted
    // annotation are automatically ignored.
    var tmpId = 0

    // The @Persisted properties are managed
    @Persisted var firstName = ""
    @Persisted var lastName = ""

    // Read-only properties are automatically ignored
    var name: String {
        return "\(firstName) \(lastName)"
    }

    // If you mix the pre-10.10 property declaration
    // syntax `@objc dynamic` with the 10.10+ @Persisted
    // annotation within a class, `@objc dynamic`
    // properties are ignored.
    @objc dynamic var email = ""
}
// :code-block-end:

// :code-block-start: ignore-a-property-objc-dynamic
class ObjectModelsObjcDynamicExamples_Person: Object {
    @objc dynamic var tmpId = 0
    @objc dynamic var firstName = ""
    @objc dynamic var lastName = ""

    // Read-only properties are automatically ignored
    var name: String {
        return "\(firstName) \(lastName)"
    }

    // Return a list of ignored property names
    override static func ignoredProperties() -> [String] {
        return ["tmpId"]
    }
}
// :code-block-end:

// :code-block-start: realm-object-enum
// Define the enum
enum ObjectModelsExamples_TaskStatusEnum: String, PersistableEnum {
    case notStarted
    case inProgress
    case complete
}

// To use the enum:
class ObjectModelsExamples_Task: Object {
    @Persisted var name: String = ""
    @Persisted var owner: String?

    // Required enum property
    @Persisted var status = ObjectModelsExamples_TaskStatusEnum.notStarted // :emphasize:

    // Optional enum property
    @Persisted var optionalTaskStatusEnumProperty: ObjectModelsExamples_TaskStatusEnum? // :emphasize:
}
// :code-block-end:

// :code-block-start: realm-object-enum-objc-dynamic
// Define the enum
@objc enum ObjectModelsObjcDynamicExamples_TaskStatusEnum: Int, RealmEnum {
    case notStarted = 1
    case inProgress = 2
    case complete = 3
}

// To use the enum:
class ObjectModelsObjcDynamicExamples_Task: Object {
    @objc dynamic var name: String = ""
    @objc dynamic var owner: String?

    // Required enum property
    @objc dynamic var status = ObjectModelsObjcDynamicExamples_TaskStatusEnum.notStarted // :emphasize:
    // Optional enum property
    let optionalTaskStatusEnumProperty = RealmProperty<ObjectModelsObjcDynamicExamples_TaskStatusEnum?>() // :emphasize:
}
// :code-block-end:

class ObjectModelsExamples_MyModel: Object {
    @Persisted var someProperty = 0
}

class ObjectModels: XCTestCase {
    func testGenericCollectionFunc() {
        // :code-block-start: generic-collection
        func operateOn<C: RealmCollection>(collection: C) {
            // Collection could be either Results or List
            print("operating on collection containing \(collection.count) objects")
        }
        // :code-block-end:
    }

    func testAnyRealmCollection() {
        // :code-block-start: any-realm-collection
        class ViewController {
        //    let collection: RealmCollection
        //                    ^
        //                    error: protocol 'RealmCollection' can only be used
        //                    as a generic constraint because it has Self or
        //                    associated type requirements
        //
        //    init<C: RealmCollection>(collection: C) where C.ElementType == ObjectModelsExamples_MyModel {
        //        self.collection = collection
        //    }

            let collection: AnyRealmCollection<ObjectModelsExamples_MyModel>

            init<C: RealmCollection>(collection: C) where C.ElementType == ObjectModelsExamples_MyModel {
                self.collection = AnyRealmCollection(collection)
            }
        }
        // :code-block-end:
    }
}

// :replace-end:
